### PR TITLE
add reCAPTCHA protection to login and password reset

### PR DESCRIPTION
**Pull request by ThrawnCA**
If reCAPTCHA keys are present in the config, then add reCAPTCHA to the login page and password reset page.

Exclude user profile editing from CAPTCHA checks since that does check authentication (if you're changing your password or a sysadmin is editing someone else's profile) but is not a behaviour we need to protect from bots.

### Features:

- [ ] includes tests covering changes
- [ ] includes updated documentation
- [X] includes user-visible changes
- [ ] includes API changes
- [ ] includes bugfix for possible backport

Please [X] all the boxes above that apply

### CODE REVIEW

File: ckan/views/user.py [hunk 1]
```diff
@@ -672,6 +673,14 @@ def _prepare(self):
 
     def post(self) -> Response:
         self._prepare()
+
+        try:
+            captcha.check_recaptcha(request)
+        except captcha.CaptchaError:
+            error_msg = _(u'Bad Captcha. Please try again.')
+            h.flash_error(error_msg)
+            return h.redirect_to(u'/user/reset')
```
Comment (smotornyuk) [inline ckan/views/user.py:682]: Use `return h.redirect_to('user.request_reset')` instead of static path to be sure that URL is correct in edge cases.

**Comment by smotornyuk** [comment]
Looks cool, only one small change of redirect required and it's ready for merge

**Comment by smotornyuk** [comment]
Oooops, just noticed a problem
Because you are using `g` inside authenticator, tests are failing
image

I can't recall any scenario where authenticator is used outside of request context, so if adding `with_request_fixture` to these 2 failing tests solves the problem, I'd suggest it as a fix

File: ckan/lib/authenticator.py [hunk 2]
```diff
@@ -24,9 +26,23 @@
...
+            # Check for a valid reCAPTCHA response
+            try:
+                client_ip_address = request.remote_addr or 'Unknown IP Address'
+                captcha.check_recaptcha_v2_base(
+                    client_ip_address,
+                    request.form.get(u'g-recaptcha-response', '')
+                )
+                return user_obj
+            except captcha.CaptchaError:
+                log.warning('Login as %r failed - failed reCAPTCHA', login)
```
[the remark this answers is not in the document]
Comment (ThrawnCA) [reply ckan/lib/authenticator.py:42]: This detector is wrong; the field contains the username, not the password.

**Pull request merged**